### PR TITLE
External project cleanup

### LINKED ISSUE #309: CIL currently failing on Travis
PR #307 only changes a few lines of code, but many of the Travis builds are failing with errors along the lines of:
```
======================================================================
ERROR: test_DataProcessor (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_DataProcessor
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/loader.py", line 428, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.5/unittest/loader.py", line 369, in _get_module_from_name
    __import__(name)
  File "/home/travis/build/CCPPETMR/SIRF-SuperBuild/sources/CCPi-Framework/Wrappers/Python/test/test_DataProcessor.py", line 21, in <module>
    from ccpi.framework import DataProcessor
  File "/home/travis/build/CCPPETMR/SIRF-SuperBuild/INSTALL/python/ccpi/framework/__init__.py", line 30, in <module>
    from .framework import DataContainer
  File "/home/travis/build/CCPPETMR/SIRF-SuperBuild/INSTALL/python/ccpi/framework/framework.py", line 47, in <module>
    cilacc = ctypes.cdll.LoadLibrary(dll)
  File "/usr/lib/python3.5/ctypes/__init__.py", line 425, in LoadLibrary
    return self._dlltype(name)
  File "/usr/lib/python3.5/ctypes/__init__.py", line 347, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcilacc.so: cannot open shared object file: No such file or directory
```
I haven't done any digging, but I suspec this is due to an upstream change to CIL?

### LINKED ISSUE #370: Code duplication in External_astra-toolbox.cmake
There's a big `IF`/`ELSE` statement in [External_astra-toolbox.cmake](https://github.com/CCPPETMR/SIRF-SuperBuild/blob/bd77f4eb40ce6842be727977d388860775c2d0d5/SuperBuild/External_astra-toolbox.cmake#L89), that depends on CUDA being present or not. I haven't looked much at it, but I'm positive this could be simplified into modifying strings dependent on the presence of CUDA, and removing ~40 lines of code. 

E.g., when CUDA is present, this is added to the `BUILD_COMMAND`: `--with-cuda=${CUDA_TOOLKIT_ROOT_DIR}`. Instead use `SET(ASTRA_CUDA_OPTIONS "--with-cuda=${CUDA_TOOLKIT_ROOT_DIR}")` such that `ASTRA_CUDA_OPTIONS` is a blank string if CUDA is not present.

### PR DESCRIPTION
**Pull request by paskino**
closes #370 
closes #309 

Additionally uses CMake macros `SetCanonicalDirectoryNames`, `SetGitTagAndRepo` and `SetExternalProjectFlags` in *all* files.

### CODE REVIEW

**Comment by paskino** [comment]
Interestingly the travis build, though listed as failed is [green](https://travis-ci.org/github/CCPPETMR/SIRF-SuperBuild/builds/669081357), at least at the previous to last commit.

File: SuperBuild/External_CCPi-Astra.cmake [hunk 1]
```diff
@@ -55,19 +52,13 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
...
-      TMP_DIR ${${proj}_TMP_DIR}
-      INSTALL_DIR ${libcilreg_Install_Dir}
+      ${${proj}_EP_ARGS_GIT}
+      ${${proj}_EP_ARGS_DIRS}
 
       CONFIGURE_COMMAND ""
       BUILD_COMMAND ""
       INSTALL_COMMAND ${CMAKE_COMMAND} -E copy_directory ${${proj}_SOURCE_DIR}/Wrappers/Python/ccpi ${PYTHON_DEST}/ccpi
-      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${libcilreg_Install_Dir}
+      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${${proj}_INSTALL_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_CCPi-Astra.cmake]:
```suggestion
      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
```

File: SuperBuild/External_CCPi-Framework.cmake [hunk 2]
```diff
@@ -62,17 +50,12 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
...
-      STAMP_DIR ${${proj}_STAMP_DIR}
-      TMP_DIR ${${proj}_TMP_DIR}
-      INSTALL_DIR ${libcilreg_Install_Dir}
+      ${${proj}_EP_ARGS_DIRS}
       # apparently this is the only way to pass environment variables to
       # external projects
       CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=${${proj}_TAG}
         ${CMAKE_COMMAND} ${${proj}_SOURCE_DIR}
-          -DCMAKE_INSTALL_PREFIX:STRING=${libcilreg_Install_Dir}
+          -DCMAKE_INSTALL_PREFIX:STRING=${${proj}_INSTALL_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_CCPi-Framework.cmake]:
```suggestion
          -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
```

File: SuperBuild/External_CCPi-FrameworkPlugins.cmake [hunk 3]
```diff
@@ -55,19 +47,13 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
...
-      TMP_DIR ${${proj}_TMP_DIR}
-      INSTALL_DIR ${libcilreg_Install_Dir}
+      ${${proj}_EP_ARGS_GIT}
+      ${${proj}_EP_ARGS_DIRS}
 
       CONFIGURE_COMMAND ""
       BUILD_COMMAND ""
       INSTALL_COMMAND ${CMAKE_COMMAND} -E copy_directory ${${proj}_SOURCE_DIR}/Wrappers/Python/ccpi ${PYTHON_DEST}/ccpi
-      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${libcilreg_Install_Dir}
+      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${${proj}_INSTALL_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_CCPi-FrameworkPlugins.cmake]:
```suggestion
      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
```

File: SuperBuild/External_CCPi-Regularisation-Toolkit.cmake [hunk 4]
```diff
@@ -79,29 +72,22 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
...
+      ${${proj}_EP_ARGS_DIRS}
       # apparently this is the only way to pass environment variables to
       # external projects
       CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=${${proj}_TAG} ${CMAKE_COMMAND} ${${proj}_SOURCE_DIR}
-        -DCMAKE_INSTALL_PREFIX=${libcilreg_Install_Dir}
-        -DBUILD_PYTHON_WRAPPER=ON -DCMAKE_BUILD_TYPE=Release
-        -DBUILD_CUDA:BOOL=${${proj}_USE_CUDA} -DCONDA_BUILD=OFF -DPYTHON_EXECUTABLE=${PYTHON_EXECUTABLE}
+        -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
+        -DBUILD_PYTHON_WRAPPER:BOOL=ON -DCMAKE_BUILD_TYPE:STRING=Release
+        -DBUILD_CUDA:BOOL=ON -DCONDA_BUILD:BOOL=OFF
```
Comment (KrisThielemans) [inline SuperBuild/External_CCPi-Regularisation-Toolkit.cmake]: CUDA stuff conflicts with @rijobro's PR I guess

File: SuperBuild/External_CCPi-Regularisation-Toolkit.cmake [hunk 5]
```diff
@@ -114,19 +100,13 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
...
-        TMP_DIR ${${proj}_TMP_DIR}
-        INSTALL_DIR ${libcilreg_Install_Dir}
+        ${${proj}_EP_ARGS_GIT}
+        ${${proj}_EP_ARGS_DIRS}
         CONFIGURE_COMMAND ""
         BUILD_COMMAND ""
         INSTALL_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=${${proj}_TAG} SRC_DIR=${${proj}_BINARY_DIR} RECIPE_DIR=${${proj}_SOURCE_DIR}/Wrappers/Python/conda-recipe PYTHON=${PYTHON_EXECUTABLE} bash ${${proj}_SOURCE_DIR}/Wrappers/Python/conda-recipe/build.sh
         CMAKE_ARGS
-           -DCMAKE_INSTALL_PREFIX=${libcilreg_Install_Dir}
+           -DCMAKE_INSTALL_PREFIX=${${proj}_INSTALL_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_CCPi-Regularisation-Toolkit.cmake]:
```suggestion
           -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
```

File: SuperBuild/External_TomoPhantom.cmake [hunk 6]
```diff
@@ -34,42 +34,26 @@ ExternalProject_Include_Dependencies(${proj} DEPENDS_VAR ${proj}_DEPENDENCIES)
...
-    TMP_DIR ${${proj}_TMP_DIR}
+    ${${proj}_EP_ARGS_GIT}
+    ${${proj}_EP_ARGS_DIRS}
+    
 
     CMAKE_ARGS
-      -DCMAKE_INSTALL_PREFIX=${TomoPhantom_Install_Dir}
-    	-DLIBRARY_DIR:PATH=${TomoPhantom_Install_Dir}/lib
-    	-DINCLUDE_DIR:PATH=${TomoPhantom_Install_Dir}/include
+      -DCMAKE_INSTALL_PREFIX=${${proj}_INSTALL_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_TomoPhantom.cmake]:
```suggestion
      -DCMAKE_INSTALL_PREFIX:PATH=${${proj}_INSTALL_DIR}
```

File: SuperBuild/External_astra-toolbox.cmake [hunk 7]
```diff
@@ -161,7 +106,25 @@ CPPFLAGS=\"-DASTRA_PYTHON -I${SUPERBUILD_INSTALL_DIR}/include -L${SUPERBUILD_INS
 
 
     endif()
+    ExternalProject_Add(${proj}
+        ${${proj}_EP_ARGS}
+        ${${proj}_EP_ARGS_GIT}
+        ${${proj}_EP_ARGS_DIRS}
+        
+        # This build is Unix specific
```
Comment (KrisThielemans) [inline SuperBuild/External_astra-toolbox.cmake]: do we check for this?

**Comment by KrisThielemans** [comment]
I see lots of trouble with the github/travis integration these days. sigh

File: SuperBuild/External_CCPi-Regularisation-Toolkit.cmake (hunk 4 above)
Comment (paskino) [reply SuperBuild/External_CCPi-Regularisation-Toolkit.cmake]: Not really. `BUILD_CUDA` is a request from the user and could be harmonised with the general build cuda option.

File: SuperBuild/External_astra-toolbox.cmake (hunk 7 above)
Comment (paskino) [reply SuperBuild/External_astra-toolbox.cmake]: We don't at the minute

File: SuperBuild/External_CCPi-Regularisation-Toolkit.cmake (hunk 4 above)
Comment (rijobro) [reply SuperBuild/External_CCPi-Regularisation-Toolkit.cmake:85]: looks like you replaced 
```
-DBUILD_CUDA:BOOL=${${proj}_USE_CUDA}
```
with
```
-DBUILD_CUDA:BOOL=ON
```
which means you now ignore the user's wishes to potentially disable CUDA.

**Comment by KrisThielemans** [comment]
I'll take care of this one.

File: SuperBuild/External_CCPi-Regularisation-Toolkit.cmake (hunk 4 above)
Comment (KrisThielemans) [reply SuperBuild/External_CCPi-Regularisation-Toolkit.cmake:85]: @paskino, maybe CUDA is required for the CCPi-Regularisation-Toolkit? if not, I'll revert to use the variable

Comment (KrisThielemans) [reply SuperBuild/External_CCPi-Regularisation-Toolkit.cmake:85]: @paskino confirmed that there's both CPU and GPU versions, so I'll reenable the bool

**Comment by KrisThielemans** [comment]
Travis will likely fail due to https://github.com/SyneRBI/SIRF/issues/637

**Pull request merged**